pkg/collector/reconcile/statefulset_test.go

```diff
@@ -0,0 +1,156 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/open-telemetry/opentelemetry-operator/api/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/internal/config"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/reconcile"
+)
+
+var logger = logf.Log.WithName("unit-tests")
+
+func TestExpectedStatefulSets(t *testing.T) {
+	// prepare
+	var prev_replicas int32 = 3
+	var new_replicas int32 = 4
+	cfg := config.New()
+	nsn := types.NamespacedName{Name: "existingstatefulset", Namespace: "default"}
+	labels := client.MatchingLabels(map[string]string{
+		"app.kubernetes.io/managed-by": "opentelemetry-operator",
+	})
+
+	// First create a StatefulSet instance in the client as the "existing" instance
+	existed := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nsn.Name,
+			Namespace: nsn.Namespace,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &prev_replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+			},
+		},
+	}
+
+	err := k8sClient.Create(context.Background(), existed)
```

Comment:
Does this test require a running k8s cluster and relevant configuration?  If so, it should probably be treated as an integration test, guarded by a build flag, and only run when an appropriate environment is known to exist.

Reply:
Yes, this is actually not a unit test. The environment setup is done in suite_test.go. I just followed the same pattern in the controller test.

Reply:
envtest brings up apiserver and etcd using binaries which get installed with kubebuilder. So basically we are not actually starting the whole cluster - only these two components